design-proposals/storage-class-migration/storage-class-migration.md

```diff
@@ -0,0 +1,276 @@
+
+However, the core idea is that it is KubeVirt responsability to handle fate of the source PVCs as it is the only component able to orchestrate and set the migration as successfull.
+
+## Design details and motivation
+
+### Motivation on the approaches
+
+From libvirt side there a multiple ways to achieve the storage migration. It is possible to simply copy the storage on the new destination and then swap the original disk wit the new one. This approach is the most perfomant as the overhead is caused only by the storage copy.
+
+A different approach is to simultaneously move the storage and live-migrate the guest. When the destination storage is not available on the host where the VM is executing, this solution is required. Costs are greater since there is a storage migration overhead as well as a live VM migration overhead.
+
+As KubeVirt is designed today and due to Kubernetes design principles of volume pod immutability, it isn't possible to hotplug and unplug new volumes while the pod is running. KubeVirt already supports volume hotplug by attaching new volume to a dummy pod and then by bound mounting the storage in `virt-launcher` pod. Sadly, this method fails to address the issue of removing the old PVCs that were expressly mentioned in the virt-launcher pod.
+Declaring all volumes hotpluggable could be a potential remedy, but we prefer to avoid it because it has been shown to be fragile for some storage providers and opaque to Kubernetes.
+
+The only workable and generic solution for this restriction appears to be live VM migration with storage. Live migration involves the creation of a new virt-launcher pod by KubeVirt that has the destination PVCs attached. The source virt-launcher pod and source PVCs are removed after the migration is finished because they are no longer needed.
+
+This migrating to a new storage provider is an exceptional operations as very resource consuming. If in the future, the k8s community will consider the volume mutability in the pod spec, then the lighter approach could be taken into consideration.
+
+### Design
+
+This feature adds an additional controller for the CRD `Storage Migration` to virt-controller. Once the StorageMigration object is created then KubeVirt will update the VMI status with the source and destination volume pair.
+If the user specifies the storage migration classes, then the destination PVCs will be created before updating the VMI status.
+
+Afterwards, the controller will create a VirtualMachineInstanceMigration to trigger the VM live migration.
+
+The controller if there are some migrated volume in the VMI status, it will replace the destination volumes by creating the target virt-launcher pod and virt-launcher will select the source volume to be migrated.
+
+Once the migration has succeeded, the controller will update firstly the migrated volume under the VMI specification. This update will consequentally trigger the update of the volume in the VM.
```

Comment:
`Once the migration completes, the controller will update the migrated volume first in the VMI spec. This will trigger the update of the volume in the VM spec`